### ISSUE TITLE
CUDA toolkit is missing a libcuda.so.1 symlink to stub lib

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#42825]
> [!TIP]
> Workaround is to use `ldflags=-Wl,--allow-shlib-undefined` in Spack 0.22.0.dev0 and above. In the future we could look to do this by default.

### Steps to reproduce

When you link a shared lib against the cuda stub lib

```
gcc -o libf.so f.o -lcuda
```

you get a `libf.so` with `DT_NEEDED` specified as `libcuda.so.1`, because that is the `SONAME` of the stub lib.

If you then link an executable against this library, it fails:

```
gcc -o app app.o -lf
```

because the _linker_ has to locate transitive deps (just like the runtime loader would), and cannot find `libcuda.so.1` because it does not exist in the stub lib dir.

```
ld: warning: libcuda.so.1, needed by libf.so, not found (try using -rpath or -rpath-link)
```

The hint about `-rpath` and `-rpath-link` does not apply, because there is no file `libcuda.so.1` anywhere.

This would be solved if the CUDA stub libs dir contained:

```
libcuda.so.1 -> libcuda.so
```

It's likely that NVIDIA folks did not consider this use case, they may have purposefully omitted this symlink to avoid that the stub lib gets used at runtime, but glanced over the fact that sometimes you need to link against a library linked against a stub lib, for which `libcuda.so.1` is required.

**Comment by robertmaynard** [comment]
Directly linking to the stub library itself avoids the issue when the SONAME don't match. We just implemented this in CMake: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9050

But yes my understanding is the symlink of `libcuda.so.1` doesn't exist in the stubs directory due to the concern that it can break runtime loading.

**Comment by haampie** [comment]
I don't think that concern outweighs the issue it creates. The stub libraries are in a separate directory, which is unlikely to be used as an rpath or pop up in LD_LIBRARY_PATH. Linking to a library that links against a stub lib however happens all the time.

A fix for CMake does not help with autotools, meson and makefile packages

**Comment by robertmaynard** [comment]
We should clarify that this only occurs on machines that don't have the CUDA driver in an implicit system path. So to trigger this behavior you need to have the CUDA toolkit installed but no driver, which generally would be a CI build machine....

**Comment by giordano** [comment]
> which generally would be a CI build machine....

Or the login node of a cluster, which doesn't exposes the drivers of the GPUs because they're only available on the compute nodes. This setup is far from being unusual.

**Comment by robertmaynard** [comment]
> > which generally would be a CI build machine....
> 
> Or the login node of a cluster, which doesn't exposes the drivers of the GPUs because they're only available on the compute nodes. This setup is far from being unusual.

Sorry, to be clear I wasnt disagreeing. The issue is common, but for people to reproduce they need to have the full context

**Comment by haampie** [comment]
It's common for login nodes, and for build bots / CI.

**Comment by haampie** [comment]
FWIW, we could in principle fix this in Spack, for example:

1. Create the symlink `<stubs dir>/libcuda.so.1 -> libcuda.so` ourselves post-install of `cuda`, and set `-Wl,-rpath-link=<stub dir>` as `LDFLAGS` in the build env of any package that depends on `cuda`.

Alternatively we could put the symlink in the stage dir or copy the entire stub lib as libcuda.so.1, whatever is supported by the file system.

Not 100% sure how reliably we can get that flag `-Wl,-rpath-link` injected though. And also have to think about the transitive case where pkg `A` links to a library from pkg `B` which links against the libcuda stub lib.

But simpler would be if `cuda` did this as part of its installation.

**Comment by giordano** [comment]
> Not 100% sure how reliably we can get that flag `-Wl,-rpath-link` injected though.

Something along these lines
```py
    def setup_dependent_build_environment(self, env, dependent_spec):
        # [...]
        libcuda = find_libraries("libcuda", root=self.prefix, shared=True, recursive=True)
        libdirs = set([os.path.dirname(lib) for lib in libcuda])
        for dir in libdirs:
            env.append_path("LIBRARY_PATH", dir)
```
should help, right?  I used that successfully to let the linker find the stub `libcuda.so` library, before realising I was still missing `libcuda.so.1`

**Comment by haampie** [comment]
Yeah, `LIBRARY_PATH` is maybe easier to set than `LDFLAGS`, cause it typically doesn't override build system defaults.

I verified that `setup_dependent_build_environment` is called when `cuda` is an indirect link dependency, so that's good.

I don't think we have a callback that lets a package modify the staging dir of its dependent, so not easy to put a `libcuda.so.1` outside of cuda's prefix.

**Comment by robertmaynard** [comment]
It should be noted that this is only needed when using `ld.bfd` in combination with `--no-allow-shlib-undefined`. So when using linkers such as gold, mold, and lld this transitive symbol requirement behavior isn't needed 
-    https://github.com/llvm/llvm-project/commit/b4744d306c0a21933d4339fa4fb497866a36ee49#diff-050be97050a41e1b5f83c19795b5e33d
- https://github.com/llvm/llvm-project/issues/51774
- https://reviews.llvm.org/D18269#378717

In those cases once you have finished creating the DSO (`B`) that links to libcuda.so you don't need to worry about ensuring a proper rpath-link line for consumers of `B`. 

I would be hestitant ( since I have CMake myopic view ) to add a `libcuda.so` symlink to the CUDA Toolkit stubs directory.  
CMake ( pre 3.28? )  will generate `-rpath` entries for every non system directory and you will end up with runtime failures due to loading the stub version ( unless you are also passing RPATH controls like [CMAKE_BUILD_WITH_INSTALL_RPATH](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_WITH_INSTALL_RPATH.html) ). These versions of CMake work currently due to the fact that the `libcuda.so.1` is missing from stubs, but I have seen this fail when devs/envs/etc add this symlink.

I would recommend placing the symlink in a separate directory that is either an implicit link dir of the compiler, or entirely separate and added to the search paths with `-rpath-link`.

Note: 
That this issue in the CUDA Toolkit isn't isolated to just libcuda.so, but also at least `libnvidia-ml.so` since it is part of the CUDA Driver and not the Toolkit.

**Comment by haampie** [comment]
You mean if CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON is used, right? I was hoping it would be smart enough to drop that dir from rpaths in FindCUDAToolkit.

**Comment by robertmaynard** [comment]
> You mean if CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON is used, right? I was hoping it would be smart enough to drop that dir from rpaths in FindCUDAToolkit.

Depends if we are talking about build or install versions of a library. Thinking about this more, I presume Spack always installs CMake projects and therfore we should talk about what the install version of libraries will have. The ` CMAKE_BUILD_WITH_INSTALL_RPATH` can be ignored in this case as it just makes the build version of libraries behave as the install version.

When installing a CMake project ( `make install` or `cmake --install`), all RPATH entries added **by CMake** during build time are dropped, and only the explicit set in the targets `INSTALL_RPATH` property are used. But if the implicit enviornment injects rpath entries via `CXXFLAGS` or other approaches those aren't removed since they are presumed to be required by the env.

So in thise case since CMake would add the `-rpath` entries via CUDAToolkit targets they would be safely removed on the install version of the library. The presumes as your correctly pointed out that `INSTALL_RPATH_USE_LINK_PATH` isn't enabled by the project ( it is disabled by default ).

**Comment by haampie** [comment]
> I presume Spack always installs

Yes, but Spack uses `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON` :)

I didn't look at it, but would still expect CMake has a way to avoid registering the stub dir even when `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON`.

**Comment by robertmaynard** [comment]
> > I presume Spack always installs
> 
> Yes, but Spack uses `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON` :)
> 
> I didn't look at it, but would still expect CMake has a way to avoid registering the stub dir even when `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON`.

There are no existing explicit controls on setting up an exclusion list for `CMAKE_INSTALL_RPATH_USE_LINK_PATH=ON`. We would need to extend `cmComputeLinkInformation` to understand an exclusion list for `CMAKE_INSTALL_RPATH_USE_LINK_PATH`.

**Comment by haampie** [comment]
OK. I think it's important for `CMAKE_INSTALL_RPATH_USE_LINK_PATH` to work, cause Spack is too high-level to know what rpaths to inject. So, I tend to agree that adding the `libcuda.so.1` symlink in the stubs dir is problematic.

Alternative solutions: 

1. Convince the linker not to resolve dependencies (there might be a good reason, but to be honest it's kinda weird)
2. Put a symlink / copy in the stage dir with `SONAME` as file name and set `LIBRARY_PATH` to it.

**Comment by haampie** [comment]
As a workaround for this issue, pass `-Wl,--allow-shlib-undefined`.

That said, I'm afraid that `ldflags` in Spack has a bug, cause the compiler wrapper does:

```sh
# Linker flags
case "$mode" in
    ld|ccld)
        extend spack_flags_list SPACK_LDFLAGS
        ;;
esac
```

which means it'll pass it to `ld` and `gcc` the same way :(

So... use `cflags=-Wl,--allow-shlib-undefined cxxflags=-Wl,--allow-shlib-undefined fflags=-Wl,--allow-shlib-undefined` as a stop gap?

**Comment by giordano** [comment]
Yesterday I was able to compile
```
'dbcsr cuda_arch=80 fflags=-Wl,--allow-shlib-undefined cxxflags=-Wl,--allow-shlib-undefined'
```
on a login node without a GPU (and without `libcuda.so.1`)

**Comment by haampie** [comment]
Submitted a PR to fix the `ldflags` situation. I've seen `cflags=...` go wrong as it would replace the defaults of a build system, effectively dropping `-O2 -g` flags, leading to unoptimized builds :sweat_smile: better to use `ldflags` only

**Comment by haampie** [comment]
Ok, you should be able to set `ldflags=-Wl,--allow-shlib-undefined` now that d946c37cbb7a41acbb2b34fc98078acc96533705 is on develop.

**Comment by polarathene** [comment]
Correct me if I'm wrong (_as I'm not using spack myself_), but isn't the unversioned `.so` file only relevant during build at link time?

Then the output binary will get it's `DT_NEEDED` entry updated with the library's `DT_SONAME` entry?

Check out this example:

```console
$ docker run --rm -it fedora:41
$ dnf config-manager addrepo --from-repofile https://developer.download.nvidia.com/compute/cuda/repos/fedora41/x86_64/cuda-fedora41.repo
$ dnf install -yq cuda-driver-devel-12-9 patchelf

$ patchelf --print-soname /usr/local/cuda-12.9/lib64/stubs/libcuda.so
libcuda.so.1

$ ls -1 /usr/local/cuda-12.9/lib64/stubs
libcuda.so

# NOTE: If /etc/ld.so.conf config would include the stubs path, this will create cache for it:
# (by default also creates a symlink for the library if the filename is not equal to the DT_SONAME entry)
$ ldconfig --verbose 2>/dev/null | grep -F 'libcuda.so'

# If it did you'll get a result here for the symlink (while verbose output is only a soname mapping):
$ ldconfig --print-cache | grep -F 'libcuda.so'

# This will resolve if you have `LD_LIBRARY_PATH` ENV for the dir, or `ld.so.cache` with a cached entry:
$ ldd /example/app | grep -F 'libcuda.so'
        libcuda.so.1 => /usr/local/cuda-12.9/lib64/stubs/libcuda.so.1 (0x00007fa3a9400000)
```

When the search path to the stub dir wasn't present, you can avoid configuring it to be detected at runtime but still use `ldconfig` command to generate expected symlink:

```console
# If it is not configured, you can generate the symlink without adding a cached entry:
$ ldd /example/app | grep -F 'libcuda.so'
        libcuda.so.1 => not found

# Symlink created by providing the directory explicitly, cache DB update prevented by `-N`:
$ ldconfig -N --verbose /usr/local/cuda-12.9/lib64/stubs 2>/dev/null | grep -F 'libcuda.so'
        libcuda.so.1 -> libcuda.so (changed)

# No cache entry:
$ ldconfig --print-cache | grep -F 'libcuda.so'

# Symlink created:
$ ls -1 /usr/local/cuda-12.9/lib64/stubs
libcuda.so
libcuda.so.1

# No search path, so this is expected:
$ ldd /example/app | grep -F 'libcuda.so'
        libcuda.so.1 => not found

# Now you can resolve as intended (actual runtime would use the correct search path instead)
$ LD_LIBRARY_PATH=/usr/local/cuda-12.9/lib64/stubs ldd /example/app | grep -F 'libcuda.so'
        libcuda.so.1 => /usr/local/cuda-12.9/lib64/stubs/libcuda.so.1 (0x00007fa3a9400000)
```

Biggest point to be mindful of there is if you want that behaviour on the same system, you need to avoid having a search path to the stub dir, and when providing an explicit one, prevent it from being added to the `ld.so.cache`.

When compiling for GCC/Clang you can use `LIBRARY_PATH` ENV, but other tooling may link without respecting that. For a rust project I noticed that was the case with Zig as the linker, and while the `cudarc` crate added common directory prefixes/suffixes for the CUDA libraries to search, if that was not the case `LIBRARY_PATH` would not work either, and like with Zig I'd have to use the `rustc` equivalent of `RUSTFLAGS='-L /path/to/lib/dir'`.

---

Hopefully that clarifies some potential knowledge gaps 😅 (_unless I misunderstood something_)

Related to CUDA specifically. If you are running an older CUDA runtime driver version but the program was built with a newer version of CUDA, you may need to use a compat package, in the official Nvidia images these are at `/usr/local/cuda/compat` where you can use `LD_LIBRARY_PATH` to use that `libcuda.so.1` instead (_eg: The output of `nvidia-smi --version` will reflect the newer CUDA runtime version is used_).

The stubs dir and unversioned `.so` files should only really be for the build environment, so the concern with the stub dir shouldn't happen in a runtime environment unless that's also populated with these build-time specific libs/dirs and that system is misconfigured :)

**Comment by giordano** [comment]
> Correct me if I'm wrong (as I'm not using spack myself), but isn't the unversioned `.so` file only relevant during build at link time?

No: when you need to link library A which links to another library B which in turn links to library C, the latter is searched by the linker _on disk_ by filename using the DT_NEEDED name stored in library B header. Not having a library with filename matching the soname (ensuring this via symlinks is fine) is incredibly dodgy because that's not how linkers work.

**Comment by polarathene** [comment]
@giordano I didn't mean that, I was referring to `-lcuda` begins search for `libcuda.so`,then from there the symlink is resolved and `DT_SONAME` is added to `DT_NEEDED` of your built binary.

So `libcuda.so.1` being the soname to link is what will be searched and required at runtime. In the nvidia images you have compat packages symlink `.so` => `.so.1` => `.so.575.57.08` (which has `libcuda.so.1` as `DT_SONAME`).

I can build without needing `libcuda.so.1` present. At runtime, `DT_NEEDED` entries will be resolved and then that must be discoverable instead of `libcuda.so` which is no longer relevant. After that anything that `libcuda.so.1` has in its `DT_NEEDED` entry is resolved.

If I've misunderstood something could you better elaborate with an example? I can provide one for what I've said if wanted.

**Comment by giordano** [comment]
> then from there the symlink is resolved and `DT_SONAME` is added to `DT_NEEDED` of your built binary.

I don't know what symlink you're talking about.  In the stubs directory there's only `libcuda.so` with soname `libcuda.so.1`, no other libcuda-looking file is around, symlink or not.

>  In the nvidia images you have compat packages symlink `.so` => `.so.1` => `.so.575.57.08` (which has `libcuda.so.1` as `DT_SONAME`).

Again, I don't see how this is relevant here: I don't use Nvidia images, no one brought them up here but you, I don't have any of those files.

**Comment by polarathene** [comment]
@giordano I mentioned it for reference. If you follow along the docker fedora container example I showed, it installs an nvidia package that provides only the stub for `libcuda.so` as is discussed here.

That file has the soname `libcuda.so.1`, and I create the symlink from that via `ldconfig` whilst avoiding adding an entry to `ld.so.conf`.

Not that you need to do such for building it, just the unversioned so file is sufficient for building, only at runtime is the linked soname filename expected to resolve.

My response covered these concerns. So far it doesn't sound like I've misunderstood anything. If you can state that actual issue that remains that'd be appreciated. The stub file is purely for convenience of satisfying the link at build without needing the full CUDA libs present that each stub represents. You'll find this explained in the CUDA 6.5 release notes when it was introduced.
